File: tools/benchmarks/bench_error.py
# Benchmark: Error handling with propagation
from dataclasses import dataclass
from typing import Union, Generic, TypeVar

T = TypeVar("T")

@dataclass
class Ok(Generic[T]):
    value: T

@dataclass
class Err:
    error: str

Result = Union[Ok[T], Err]

def parse_age(s: str) -> Result:
    try:
        n = int(s)
    except ValueError:
        return Err(f"Not a number: {s}")
    if n < 0:   return Err(f"Age cannot be negative: {n}")
    if n > 150: return Err(f"Unlikely age: {n}")
    return Ok(n)

@dataclass
class User:
    name: str
    age:  int

def validate_user(name: str, age_str: str) -> Result:
    if not name: return Err("Name cannot be empty")
    result = parse_age(age_str)
    if isinstance(result, Err): return result
    return Ok(User(name=name, age=result.value))

def process(name: str, age_str: str) -> str:
    result = validate_user(name, age_str)
    if isinstance(result, Ok):
        return f"Valid: {result.value.name}, age {result.value.age}"
    return f"Error: {result.error}"

print(process("Alice", "30"))
print(process("Bob",   "abc"))
print(process("",      "25"))
print(process("Diana", "-5"))
print(process("Eve",   "200"))

File: tools/benchmarks/bench_error.ts
// Benchmark: Error handling with propagation
type Result<T> =
  | { ok: true;  value: T }
  | { ok: false; error: string };

function parseAge(s: string): Result<number> {
  const n = parseInt(s, 10);
  if (isNaN(n))  return { ok: false, error: `Not a number: ${s}` };
  if (n < 0)     return { ok: false, error: `Age cannot be negative: ${n}` };
  if (n > 150)   return { ok: false, error: `Unlikely age: ${n}` };
  return { ok: true, value: n };
}

interface User { name: string; age: number; }

function validateUser(name: string, ageStr: string): Result<User> {
  if (name.length === 0) return { ok: false, error: "Name cannot be empty" };
  const ageResult = parseAge(ageStr);
  if (!ageResult.ok) return ageResult;
  return { ok: true, value: { name, age: ageResult.value } };
}

function process(name: string, ageStr: string): string {
  const result = validateUser(name, ageStr);
  if (result.ok) return `Valid: ${result.value.name}, age ${result.value.age}`;
  return `Error: ${result.error}`;
}

console.log(process("Alice", "30"));
console.log(process("Bob",   "abc"));
console.log(process("",      "25"));
console.log(process("Diana", "-5"));
console.log(process("Eve",   "200"));
